table/block_based_table_builder.cc

```diff
@@ -500,11 +500,16 @@ void BlockBasedTableBuilder::Add(const Slice& key, const Slice& value) {
   if (!ok()) return;
   ValueType value_type = ExtractValueType(key);
   if (IsValueType(value_type)) {
-#ifndef NDEBUG
     if (r->props.num_entries > r->props.num_range_deletions) {
-      assert(r->internal_comparator.Compare(key, Slice(r->last_key)) > 0);
+      if (r->internal_comparator.Compare(key, Slice(r->last_key)) <= 0) {
+        // We were about to insert keys out of order. Abort.
+        ROCKS_LOG_ERROR(r->ioptions.info_log,
+                        "Out-of-order key insertion into block based table");
+        r->status =
+            Status::Corruption("Out-of-order key insertion into table");
```

Comment:
@ajkr Is setting `r->status` sufficient? Or should we fatal error here?

Reply:
I think `r->status` should be fine. It should cause the table to be thrown away and the DB to enter read only mode. RocksDB's philosophy is more around returning errors than fatal erroring since it's supposed to be embedded in arbitrary applications that might not want to crash.